staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -15,58 +15,84 @@
 package model
 
 import (
-	"k8s.io/kube-openapi/pkg/validation/spec"
 	"strings"
 
 	"gopkg.in/yaml.v3"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
 )
 
 // SchemaDeclTypes constructs a top-down set of DeclType instances whose name is derived from the root
 // type name provided on the call, if not set to a custom type.
-func SchemaDeclTypes(s *spec.Schema, maybeRootType string) (*DeclType, map[string]*DeclType) {
+func SchemaDeclTypes(s *schema.Structural, maybeRootType string) (*DeclType, map[string]*DeclType) {
 	root := SchemaDeclType(s).MaybeAssignTypeName(maybeRootType)
 	types := FieldTypeMap(maybeRootType, root)
 	return root, types
 }
 
 // SchemaDeclType returns the CEL Policy Templates type name associated with the schema element.
-func SchemaDeclType(s *spec.Schema) *DeclType {
-	if len(s.Type) < 1 {
-		return NewObjectTypeRef("*error*")
+func SchemaDeclType(s *schema.Structural) *DeclType {
+	declType, found := openAPISchemaTypes[s.Type]
+	if !found {
+		return nil
 	}
-	for _, schemaType := range s.Type {
-		declType, found := openAPISchemaTypes[schemaType]
-		if !found {
-			continue
+	if s.XIntOrString {
+		return NewObjectType("intOrString", map[string]*DeclField{
+			"strVal": {Name: "strVal", Type: StringType},
+			"intVal": {Name: "intVal", Type: IntType},
+		})
+	}
+	if s.XEmbeddedResource {
+		return NewObjectType("embedded", map[string]*DeclField{
+			"kind":       {Name: "kind", Type: StringType},
+			"apiVersion": {Name: "apiVersion", Type: StringType},
+			"metadata": {Name: "metadata", Type: NewObjectType("metadata", map[string]*DeclField{
+				"name":         {Name: "name", Type: StringType},
+				"generateName": {Name: "generateName", Type: StringType, Required: false},
+			})},
+			"spec":   {Name: "spec", Type: DynType},
+			"status": {Name: "status", Type: DynType, Required: false},
+		})
+	}
+	if s.XPreserveUnknownFields {
+		return DynType
```

Comment:
this is not completely correct I think. You can have XPreserveUnknownFields despite another type, especially types like lists or map. I.e. you can partially specify a schema that way, but stop pruning in a subtree of the ast.

Reply:
We might be up against the limits of CEL here. DynType here effectively defers all type checking to runtime. My preference would that we have CEL perform static type checking on known fields but then allow all other fields through for runtime type checking, but I don't think that's possible (cc @TristonianJones).

Note that for runtime type checking, developers can still use field selection `preserveUnknownFieldsProp.something` and `has` checks, they just won't get a compilation failure if they mistype a field name.

Reply:
I've been thinking about this one some more. How about:

We don't allow validation rules on the contents of any unknown data.

This would apply to all unknown fields in "preserve unknown fields" "embedded".

This means we only validate fields that we actually have schema information about, which actually seems pretty consistent. For example, if a field is embedded, we don't run  any of its openapi validation rules do we?

This would be easy to implement too

Reply:
I've implemented this, PTAL.

Reply:
am little undecided which was is better. I can see use for checking something on unknown data, also because it might be that OpenAPI cannot express what the preserve-unknown-fields value should be and CEL could. On the other hand, this opens door for "misuse" by defining esoteric types (e.g. everything with a prime number of fields is allowed).